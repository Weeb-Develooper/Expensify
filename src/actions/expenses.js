import { v1 as uuidv1 } from "uuid";

//ADD EXPENSE
export const addExpense = ({
  description = "",
  note = "",
  amount = 0,
  createdAt = 0,
} = {}) => {
  return {
    type: "ADD_EXPENSE",
    expense: {
      id: uuidv1(),
      description,
      note,
      amount,
      createdAt,
    },
  };
};

//REMOVE EXPENSE
export const removeExpense = ({ id } = {}) => {
  return {
    type: "REMOVE_EXPENSE",
    id,
  };
};

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
  type: "EDIT_EXPENSE",
  id,
  updates,
});
