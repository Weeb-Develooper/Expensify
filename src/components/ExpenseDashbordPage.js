import React from "react";
import ExpenseList from "./ExpenseList";
import ExpenseListFilters from "./ExpenseListFilters";

const ExpenseDashbordPage = () => {
  return (
    <div>
      <ExpenseListFilters />
      <ExpenseList />
    </div>
  );
};

export default ExpenseDashbordPage;
