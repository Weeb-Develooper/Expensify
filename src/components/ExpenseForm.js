import React from "react";
import moment from "moment";
import { SingleDatePicker } from "react-dates";
import "react-dates/lib/css/_datepicker.css";

// const now = moment();
// console.log(now.format("MMM Do, YYYY"));

class ExpenseForm extends React.Component {
  //using the constructor syntax because some props are being passed down from the edit expense page
  constructor(props) {
    super(props);

    this.state = {
      description: props.expense ? props.expense.description : "",
      note: props.expense ? props.expense.note : "",
      //converting the amount to cents and then to string
      amount: props.expense ? (props.expense.amount / 100).toString() : "",
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calendarFocused: false,
      error: "",
    };
  }

  onDescriptionChange = (e) => {
    const description = e.target.value;
    //implicitly returning an object so its wrapped in brackets
    this.setState(() => ({ description }));
  };

  onNoteChange = (e) => {
    const note = e.target.value;

    this.setState(() => {
      return {
        note: note,
      };
    });
  };

  onAmountChange = (e) => {
    const amount = e.target.value;

    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  };

  onDateChange = (createdAt) => {
    if (createdAt) {
      this.setState(() => ({ createdAt }));
    }
  };

  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calendarFocused: focused }));
  };

  onSubmit = (e) => {
    e.preventDefault();

    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({ error: "Description and amount are required" }));
    } else {
      this.setState(() => ({ error: "" }));
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10) * 100,
        //valueOf converts the date to timestamps
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note,
      });
    }
  };

  render() {
    return (
      <div>
        {this.state.error != "" && <p>{this.state.error}</p>}
        <form onSubmit={this.onSubmit}>
          <input
            type="text"
            placeholder="Description"
            autoFocus
            value={this.state.description}
            onChange={this.onDescriptionChange}
          />
          {/*to validate the input so that only curency lik numbers are accepted */}
          <input
            type="text"
            placeholder="Amount"
            value={this.state.amount}
            onChange={this.onAmountChange}
          />

          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calendarFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
          />

          <textarea
            placeholder="Add a note for your expense (optional)"
            value={this.state.note}
            onChange={this.onNoteChange}
          />
          <button>Add Expense</button>
        </form>
      </div>
    );
  }
}

export default ExpenseForm;
