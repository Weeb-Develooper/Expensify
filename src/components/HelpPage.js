import React from "react";

const HelpPage = () => {
  return <div>This is from the help page</div>;
};

export default HelpPage;
