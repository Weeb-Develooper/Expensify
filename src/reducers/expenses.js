//DEFAULT EXPENSE STATE
const expensesReducersDefaultState = [];

const expensesReducer = (state = expensesReducersDefaultState, action) => {
  switch (action.type) {
    case "ADD_EXPENSE":
      return [...state, action.expense];
    case "REMOVE_EXPENSE":
      return state.filter((expense) => {
        //if the expression returns true the item would be kept if not it would be removed
        return expense.id !== action.id;
      });
    case "EDIT_EXPENSE":
      return state.map((expense) => {
        if (expense.id === action.id) {
          return {
            ...expense,
            ...action.updates,
          };
        } else {
          return expense;
        }
      });
    default:
      return state;
  }
};

export default expensesReducer;
