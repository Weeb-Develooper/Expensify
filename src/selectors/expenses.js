import moment from "moment";

// Get visible expenses
const getVisibleExpenses = (expenses, { text, sortBy, startDate, endDate }) => {
  return expenses
    .filter((expense) => {
      //convert the createdAt time stamp to a moment date
      const createdAtMoment = moment(expense.createdAt);
      //checks if there is a start date and if it the same or before the moment date above
      const startDateMatch = startDate
        ? startDate.isSameOrBefore(createdAtMoment, "day")
        : true;
      //checks if there is a end date and if it the same or after the moment date above
      const endDateMatch = endDate
        ? endDate.isSameOrAfter(createdAtMoment, "day")
        : true;
      const textMatch = expense.description
        .toLowerCase()
        .includes(text.toLowerCase());

      return startDateMatch && endDateMatch && textMatch;
    })
    .sort((a, b) => {
      if (sortBy === "date") {
        return a.createdAt < b.createdAt ? 1 : -1;
      } else if (sortBy === "amount") {
        return a.amount < b.amount ? 1 : -1;
      }
    });
};

export default getVisibleExpenses;
