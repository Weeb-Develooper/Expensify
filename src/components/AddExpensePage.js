import React from "react";
import ExpenseForm from "./ExpenseForm";
import { connect } from "react-redux";
import { addExpense } from "../actions/expenses";

const AddExpensePage = (props) => {
  return (
    <div>
      <h2>Add Expense</h2>
      <ExpenseForm
        onSubmit={(expenses) => {
          props.dispatch(addExpense(expenses));
          props.history.push("/");
        }}
      />
    </div>
  );
};

export default connect()(AddExpensePage);
