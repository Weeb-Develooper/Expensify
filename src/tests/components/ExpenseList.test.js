import { ExpenseList } from "../../components/ExpenseList";
import { shallow } from "enzyme";
import React from "react";
import expenses from "../fixtures/expenses";

test("should render expense list with expenses", () => {
  const wrapper = shallow(<ExpenseList expenses={expenses} />);
  expect(wrapper).toMatchSnapshot();
});

test("should render expense list with empty expenses", () => {
  const wrapper = shallow(<ExpenseList expenses={[]} />);
  expect(wrapper).toMatchSnapshot();
});
