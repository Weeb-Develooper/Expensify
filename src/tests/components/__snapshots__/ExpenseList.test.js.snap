// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render expense list with empty expenses 1`] = `
<div>
  <p>
    No Expenses
  </p>
</div>
`;

exports[`should render expense list with expenses 1`] = `
<div>
  <ExpenseListItem
    amount={12333}
    createdAt={0}
    description="Gum"
    id="123"
    key="123"
    note=""
  />
  <ExpenseListItem
    amount={109500}
    createdAt={-345600000}
    description="rent"
    id="345"
    key="345"
    note=""
  />
  <ExpenseListItem
    amount={4500}
    createdAt={345600000}
    description="card"
    id="678"
    key="678"
    note=""
  />
</div>
`;
